File: ambari-agent/src/main/java/org/apache/ambari/tools/hadoop/ManagedDependencyClient.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ambari.tools.hadoop;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FSDataInputStream;
import org.apache.hadoop.fs.FSDataOutputStream;
import org.apache.hadoop.fs.FileStatus;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.fs.permission.AclEntry;
import org.apache.hadoop.fs.permission.FsPermission;
import org.apache.hadoop.security.AccessControlException;
import org.apache.hadoop.security.UserGroupInformation;
import org.apache.hadoop.util.VersionInfo;

/** Stateless client observations. The Agent journal and Server own workflow state. */
public final class ManagedDependencyClient {
  private static final ObjectMapper JSON = new ObjectMapper();
  static final int PROBE_BYTES = 128 * 1024 + 1;
  private static final int MAX_MARKER_BYTES = 8192;

  private ManagedDependencyClient() {
  }

  public static void main(String[] args) throws Exception {
    if (args.length != 4) {
      throw new IllegalArgumentException("Expected operation, configuration directory, identity, payload");
    }
    String operation = args[0];
    JsonNode identity = JSON.readTree(args[2]);
    validateIdentity(identity);
    JsonNode payload = JSON.readTree(args[3]);
    Map<String, Object> response = new LinkedHashMap<>();
    response.put("schemaVersion", 1);
    response.put("identity", identity);
    response.put("operation", operation);
    try {
      Map<String, Object> result;
      if ("VERSION".equals(operation)) {
        String kind = requiredText(payload, "kind");
        // HBase is supplied by the installed client, not bundled into the Agent.
        String version = "HADOOP_CLIENT".equals(kind) ? VersionInfo.getVersion()
            : "HBASE_CLIENT".equals(kind)
                ? (String) Class.forName("org.apache.hadoop.hbase.util.VersionInfo")
                    .getMethod("getVersion").invoke(null)
                : null;
        if (version == null) {
          throw new IllegalArgumentException("Unsupported client software kind");
        }
        result = Map.of("kind", kind, "version", version);
      } else {
        Configuration config = new Configuration(false);
        config.addResource(new Path(args[1], "core-site.xml"));
        config.addResource(new Path(args[1], "hdfs-site.xml"));
        UserGroupInformation.setConfiguration(config);
        URI uri = URI.create(requiredText(payload, "path"));
        if (!"hdfs".equals(uri.getScheme()) || uri.getAuthority() == null
            || uri.getQuery() != null || uri.getFragment() != null || uri.getUserInfo() != null) {
          throw new IllegalArgumentException("Expected an explicit HDFS authority");
        }
        try (FileSystem fs = FileSystem.newInstance(uri, config)) {
          result = execute(fs, operation, identity, payload);
        }
      }
      response.put("status", "SUCCEEDED");
      response.put("result", result);
    } catch (ProbeFailure failure) {
      response.put("status", "FAILED");
      response.put("errorCode", failure.code);
    } catch (AccessControlException failure) {
      response.put("status", "FAILED");
      response.put("errorCode", "DEPENDENCY_AUTHORIZATION_FAILED");
    } catch (Exception failure) {
      response.put("status", "FAILED");
      response.put("errorCode", "DEPENDENCY_CLIENT_OBSERVATION_FAILED");
    }
    System.out.println(JSON.writeValueAsString(response));
  }

  static Map<String, Object> execute(FileSystem fs, String operation,
      JsonNode identity, JsonNode payload) throws Exception {
    Path path = new Path(requiredText(payload, "path"));
    switch (operation) {
      case "STAT":
        try {
          FileStatus status = fs.getFileLinkStatus(path);
          return Map.of("exists", true, "type", status.isSymlink() ? "SYMLINK"
                  : status.isDirectory() ? "DIRECTORY" : "FILE",
              "owner", status.getOwner(), "group", status.getGroup(),
              "mode", (int) status.getPermission().toShort());
        } catch (FileNotFoundException absent) {
          return Map.of("exists", false);
        }
      case "ACL":
        List<Map<String, Object>> entries = new ArrayList<>();
        for (AclEntry entry : fs.getAclStatus(path).getEntries()) {
          entries.add(Map.of("scope", entry.getScope().name(), "type", entry.getType().name(),
              "name", entry.getName() == null ? "" : entry.getName(),
              "permission", entry.getPermission().ordinal()));
        }
        return Map.of("entries", entries);
      case "COUNT":
        // Listing only the first child avoids a recursive traversal of provider data.
        return Map.of("empty", !fs.listStatusIterator(path).hasNext());
      case "MKDIR":
        requireMutationPath(path, identity, true);
        if (!fs.getFileStatus(path.getParent()).isDirectory()) {
          throw new IOException("Managed parent is not a directory");
        }
        FsPermission permission = permission(payload);
        fs.getConf().set("fs.permissions.umask-mode", "000");
        if (!fs.mkdirs(path, permission)) {
          throw new IOException("Managed directory creation failed");
        }
        return Map.of("created", true);
      case "SET_OWNER":
        requireMutationPath(path, identity, false);
        fs.setOwner(path, requiredText(payload, "owner"), requiredText(payload, "group"));
        return Map.of("applied", true);
      case "SET_PERMISSION":
        requireMutationPath(path, identity, false);
        fs.setPermission(path, permission(payload));
        return Map.of("applied", true);
      case "READ_MARKER":
        requireMarkerPath(path, identity);
        try (FSDataInputStream input = fs.open(path)) {
          byte[] bytes = input.readNBytes(MAX_MARKER_BYTES + 1);
          if (bytes.length > MAX_MARKER_BYTES) {
            throw new IOException("Oversized binding marker");
          }
          JsonNode marker = JSON.readTree(bytes);
          if (!marker.isObject()) {
            throw new IOException("Invalid binding marker");
          }
          return Map.of("marker", marker);
        }
      case "WRITE_MARKER":
        requireMarkerPath(path, identity);
        byte[] marker = JSON.writeValueAsBytes(payload.required("marker"));
        if (marker.length > MAX_MARKER_BYTES || !payload.get("marker").isObject()) {
          throw new IOException("Invalid binding marker");
        }
        try (FSDataOutputStream output = fs.create(path, new FsPermission((short) 0600),
            false, 4096, fs.getDefaultReplication(path), fs.getDefaultBlockSize(path), null)) {
          output.write(marker);
        }
        return Map.of("written", true);
      case "PROBE":
        requireMutationPath(path, identity, false);
        if (!path.toUri().getPath().endsWith("/root")) {
          throw new IllegalArgumentException("Probe must be inside the binding root");
        }
        return probe(fs, path, identity);
      default:
        throw new IllegalArgumentException("Unsupported client operation");
    }
  }

  private static Map<String, Object> probe(FileSystem fs, Path root, JsonNode identity)
      throws Exception {
    Path path = new Path(root, ".ambari-managed-probe-" + requiredText(identity, "operationId"));
    byte[] digest = MessageDigest.getInstance("SHA-256")
        .digest(requiredText(identity, "immutableRequestHash").getBytes(StandardCharsets.UTF_8));
    byte[] expected = new byte[PROBE_BYTES];
    for (int i = 0; i < expected.length; i++) {
      expected[i] = digest[i % digest.length];
    }
    boolean owned = false;
    try {
      if (!fs.exists(path)) {
        try (FSDataOutputStream output = fs.create(path, false)) {
          owned = true;
          output.write(expected);
        }
      }
      byte[] observed;
      try (FSDataInputStream input = fs.open(path)) {
        observed = input.readNBytes(PROBE_BYTES + 1);
      }
      if (!Arrays.equals(observed, expected)) {
        throw new ProbeFailure("DEPENDENCY_DATA_INTEGRITY_FAILED");
      }
      // Only an exact replay's verified content authorizes cleaning a preexisting probe.
      owned = true;
    } catch (ProbeFailure failure) {
      throw failure;
    } catch (AccessControlException failure) {
      throw failure;
    } catch (IOException failure) {
      throw new ProbeFailure("DEPENDENCY_DATANODE_READ_WRITE_FAILED");
    } finally {
      if (owned) {
        try {
          if (!fs.delete(path, false)) {
            throw new IOException("Probe was not deleted");
          }
        } catch (IOException failure) {
          throw new ProbeFailure("DEPENDENCY_PROBE_CLEANUP_FAILED");
        }
      }
    }
    return Map.of("readWriteVerified", true, "bytes", PROBE_BYTES, "cleaned", true);
  }

  private static void requireMutationPath(Path path, JsonNode identity, boolean parents) {
    String binding = requiredText(identity, "bindingId");
    String actual = path.toUri().getPath();
    String base = "/apps/ambari-managed/hbase/" + binding;
    if (actual.equals(base) || actual.equals(base + "/root") || actual.equals(base + "/wal")
        || (parents && List.of("/apps", "/apps/ambari-managed",
            "/apps/ambari-managed/hbase", "/apps/ambari-managed/.bindings").contains(actual))) {
      return;
    }
    throw new IllegalArgumentException("Path is outside the exact managed namespace");
  }

  private static void requireMarkerPath(Path path, JsonNode identity) {
    if (!path.toUri().getPath().equals("/apps/ambari-managed/.bindings/"
        + requiredText(identity, "bindingId") + ".json")) {
      throw new IllegalArgumentException("Path is outside the exact binding marker");
    }
  }

  private static FsPermission permission(JsonNode payload) {
    JsonNode mode = payload.required("mode");
    if (!mode.isInt() || mode.intValue() < 0 || mode.intValue() > 0777) {
      throw new IllegalArgumentException("Invalid permission bits");
    }
    return new FsPermission((short) mode.intValue());
  }

  private static void validateIdentity(JsonNode identity) {
    for (String name : List.of("bindingId", "operationId")) {
      String value = requiredText(identity, name);
      if (!UUID.fromString(value).toString().equals(value)) {
        throw new IllegalArgumentException("Invalid operation identity");
      }
    }
    if (!identity.path("epoch").isIntegralNumber() || identity.get("epoch").longValue() <= 0
        || !identity.path("immutableRequestHash").asText().matches("sha256:[0-9a-f]{64}")) {
      throw new IllegalArgumentException("Invalid operation lineage");
    }
  }

  private static String requiredText(JsonNode node, String name) {
    JsonNode value = node.required(name);
    if (!value.isTextual() || value.textValue().isEmpty()) {
      throw new IllegalArgumentException("Missing text field");
    }
    return value.textValue();
  }

  private static final class ProbeFailure extends IOException {
    private final String code;

    private ProbeFailure(String code) {
      this.code = code;
    }
  }
}

File: ambari-common/src/main/python/resource_management/libraries/functions/managed_dependency_client.py
#!/usr/bin/env python3
"""
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import json
import subprocess

from resource_management.core.signal_utils import TerminateStrategy
from resource_management.libraries.functions.managed_dependency import ManagedDependencyFailure


CLIENT_HELPER_JAR = "/var/lib/ambari-agent/tools/zkmigrator.jar"
CLIENT_HELPER_CLASS = "org.apache.ambari.tools.hadoop.ManagedDependencyClient"
ERROR_CODES = {
  "DEPENDENCY_AUTHORIZATION_FAILED",
  "DEPENDENCY_DATA_INTEGRITY_FAILED",
  "DEPENDENCY_DATANODE_READ_WRITE_FAILED",
  "DEPENDENCY_PROBE_CLEANUP_FAILED",
  "DEPENDENCY_CLIENT_OBSERVATION_FAILED",
}
RESULT_FIELDS = {
  "VERSION": {"kind": str, "version": str},
  "ACL": {"entries": list},
  "COUNT": {"empty": bool},
  "MKDIR": {"created": bool},
  "SET_OWNER": {"applied": bool},
  "SET_PERMISSION": {"applied": bool},
  "READ_MARKER": {"marker": dict},
  "WRITE_MARKER": {"written": bool},
  "PROBE": {"readWriteVerified": bool, "bytes": int, "cleaned": bool},
}


def observe_client(call, executable, config_dir, command, operation, payload,
                   user, environment, timeout=60, error_code="DEPENDENCY_CLIENT_OBSERVATION_FAILED"):
  identity = dict(command.envelope)
  arguments = (executable, "--config", config_dir, CLIENT_HELPER_CLASS,
               operation, config_dir, json.dumps(identity, sort_keys=True),
               json.dumps(payload, sort_keys=True))
  # Use the selected distribution's launcher so its SDK precedes the Agent helper.
  environment = {**(environment or {}), "HADOOP_CLASSPATH": CLIENT_HELPER_JAR,
                 "HBASE_CLASSPATH": CLIENT_HELPER_JAR}
  try:
    code, output = call(arguments, user=user, env=environment, stderr=subprocess.PIPE,
                        timeout=timeout, timeout_kill_strategy=TerminateStrategy.KILL_PROCESS_GROUP,
                        shell=False, quiet=True)[:2]
    if code != 0 or not isinstance(output, str) or len(output.encode("utf-8")) > 32 * 1024:
      raise ValueError("Invalid client process result")
    result = json.loads(output)
    if (not isinstance(result, dict) or type(result.get("schemaVersion")) is not int
        or result["schemaVersion"] != 1 or json.dumps(result.get("identity"), sort_keys=True) != json.dumps(identity, sort_keys=True)
        or result.get("operation") != operation):
      raise ValueError("Invalid client response identity or schema")
    if result.get("status") == "FAILED":
      failure_code = result.get("errorCode")
      if failure_code not in ERROR_CODES:
        raise ValueError("Invalid client error code")
      raise ManagedDependencyFailure(
        error_code if failure_code == "DEPENDENCY_CLIENT_OBSERVATION_FAILED" else failure_code,
        "The managed dependency client API could not complete the observation")
    if result.get("status") != "SUCCEEDED" or "errorCode" in result:
      raise ValueError("Invalid client response status")
    facts = result.get("result")
    fields = RESULT_FIELDS.get(operation)
    if operation == "STAT":
      if not isinstance(facts, dict) or type(facts.get("exists")) is not bool:
        raise ValueError("Invalid client file status")
      fields = {"exists": bool}
      if facts["exists"]:
        fields = {**fields, "type": str, "owner": str, "group": str, "mode": int}
    if fields is None or not isinstance(facts, dict) or set(facts) != set(fields):
      raise ValueError("Invalid client observation fields")
    if any(type(facts[key]) is not expected for key, expected in fields.items()):
      raise ValueError("Invalid client observation types")
    if operation == "STAT" and facts["exists"]:
      if facts["type"] not in {"DIRECTORY", "FILE", "SYMLINK"} or not 0 <= facts["mode"] <= 0o7777:
        raise ValueError("Invalid file type or permission bits")
    if operation == "ACL":
      for entry in facts["entries"]:
        if (not isinstance(entry, dict) or set(entry) != {"scope", "type", "name", "permission"}
            or entry["scope"] not in {"ACCESS", "DEFAULT"}
            or entry["type"] not in {"USER", "GROUP", "MASK", "OTHER"}
            or type(entry["name"]) is not str or type(entry["permission"]) is not int
            or not 0 <= entry["permission"] <= 7):
          raise ValueError("Invalid typed ACL entry")
    return facts
  except ManagedDependencyFailure:
    raise
  except Exception as error:
    raise ManagedDependencyFailure(error_code,
      "The managed dependency client returned no valid structured observation") from error
